stellargraph/core/element_data.py

```diff
@@ -279,8 +281,10 @@ def features(self, type_name, id_ilocs) -> np.ndarray:
             raise ValueError("unknown IDs")
 
         try:
-            return self._features[type_name][feature_ilocs, :]
-        except IndexError:
+            return tf.nn.embedding_lookup(
+                self._features[type_name], feature_ilocs.astype(int),
+            )
+        except InvalidArgumentError:
```

Comment:
I think `tf.nn.embedding_lookup` may not throw this error when executing on a GPU, e.g. https://www.tensorflow.org/api_docs/python/tf/compat/v1/nn/embedding_lookup

> validate_indices: DEPRECATED. If this operation is assigned to CPU, values in indices are always validated to be within range. If assigned to GPU, out-of-bound indices result in safe but unspecified behavior, which may include raising an error.

Although the non-compat version https://www.tensorflow.org/api_docs/python/tf/nn/embedding_lookup doesn't discuss this at all.

We might have to switch to manual verification, like:

```python
iloc_range = self._type_element_ilocs[type_name
feature_ilocs = id_ilocs - iloc_range.start

if (feature_ilocs < 0).any() or (feature_ilocs >= len(iloc_range)).any():
    raise ValueError("unknown IDs")

return tf.nn.embedding_lookup(...)
```

Hopefully this doesn't destroy the wins we get from using `embedding_lookup`.

I guess we could potentially switch the subtraction, validation and type casting logic to use tensorflow too, maybe as a `tf.function`? (maybe this whole function should be a `tf.function`?)

It would be good to report the results of `test_benchmark_get_features` (in `test_graph.py`) before and after each refactoring. In the best case, we'd report the progress on GPU as well as on CPU.

Reply:
good news for us, if run on a gpu it will return zero tensors with the right shape for invalid and negative indices. so we don't need to change the verification here.

I'm gonna look at using a `tf.function` for this

Reply:
At least as it is, this function is always running on CPU and so will always return an exception.

Reply:
Right, but that's only because we're forcing it to run on CPU. I'll move this discussion to the main PR.